src/ray/raylet/node_manager.cc

```diff
@@ -1493,38 +1535,67 @@ void NodeManager::ProcessSubmitTaskMessage(const uint8_t *message_data) {
   RAY_CHECK(new_scheduler_enabled_);
-  while (!tasks_to_dispatch_.empty()) {
+
+  // Check every task in task_to_dispatch queue to see
+  // whether it can be dispatched and ran. This avoids head-of-line
+  // blocking where a task which cannot be dispatched because
+  // there are not enough available resources blocks other
+  // tasks from being dispatched.
+  for (size_t queue_size = tasks_to_dispatch_.size(); queue_size > 0; queue_size--) {
     auto task = tasks_to_dispatch_.front();
     auto reply = task.first;
     auto spec = task.second.GetTaskSpecification();
+    tasks_to_dispatch_.pop_front();
+
     std::shared_ptr<Worker> worker = worker_pool_.PopWorker(spec);
-    if (worker == nullptr) {
+    if (!worker) {
+      // No worker available to schedule this task.
+      // Put the task back in the dispatch queue.
+      tasks_to_dispatch_.push_front(task);
       return;
     }
 
-    bool schedulable = new_resource_scheduler_->SubtractNodeAvailableResources(
-        self_node_id_.Binary(), spec.GetRequiredResources().GetResourceMap());
+    std::shared_ptr<TaskResourceInstances> allocated_instances(
+        new TaskResourceInstances());
+    bool schedulable = new_resource_scheduler_->AllocateLocalTaskResources(
+        spec.GetRequiredResources().GetResourceMap(), allocated_instances);
     if (!schedulable) {
-      return;
+      // Not enough resources to schedule this task.
+      // Put it back at the end of the dispatch queue.
+      tasks_to_dispatch_.push_back(task);
+      worker_pool_.PushWorker(worker);
+      // Try next task in the dispatch queue.
+      continue;
     }
-    // Handle the allocation to specific resource IDs.
-    auto acquired_resources =
-        local_available_resources_.Acquire(spec.GetRequiredResources());
-    cluster_resource_map_[self_node_id_].Acquire(spec.GetRequiredResources());
+    worker->SetOwnerAddress(spec.CallerAddress());
     if (spec.IsActorCreationTask()) {
-      worker->SetLifetimeResourceIds(acquired_resources);
+      worker->SetLifetimeAllocatedInstances(allocated_instances);
     } else {
-      worker->SetTaskResourceIds(acquired_resources);
+      worker->SetAllocatedInstances(allocated_instances);
     }
+    worker->AssignTaskId(spec.TaskId());
+    worker->AssignJobId(spec.JobId());
+    worker->SetAssignedTask(task.second);
 
     reply(worker, ClientID::Nil(), "", -1);
-    tasks_to_dispatch_.pop_front();
   }
 }
 
 void NodeManager::NewSchedulerSchedulePendingTasks() {
   RAY_CHECK(new_scheduler_enabled_);
-  while (!tasks_to_schedule_.empty()) {
+  size_t queue_size = tasks_to_schedule_.size();
+
+  // Check every task in task_to_schedule queue to see
+  // whether it can be scheduled. This avoids head-of-line
+  // blocking where a task which cannot be scheduled because
+  // there are not enough available resources blocks other
+  // tasks from being scheduled.
+  while (queue_size > 0) {
+    if (queue_size == 0) {
```

Comment:
@istoica @edoakes this if is invalid, because the condition of the outer while is `queue_size > 0`, can you fix this? thanks.

Reply:
cc @wuisawesome